skbio/sequence/_substitution.py

```diff
@@ -0,0 +1,518 @@
+    def by_name(cls, name):
+        """Load a pre-defined substitution matrix by its name.
+
+        Parameters
+        ----------
+        name : str
+            Name of the substitution matrix.
+
+        Returns
+        -------
+        SubstitutionMatrix
+            Named substitution matrix.
+
+        Raises
+        ------
+        ValueError
+            If named substitution matrix does not exist.
+
+        See Also
+        --------
+        get_names
+
+        Notes
+        -----
+        Available substitution matrix names can be obtained by ``get_names``.
+        Currently, the following names are supported:
+
+        - `NUC.4.4` (a.k.a. DNAfull): A nucleotide substitution matrix covering
+          all definite and degenerate nucleotides.
+
+        - Point Accepted Mutation (PAM) [1]_: A set of amino acid substitution
+          matrices, including `PAM30`, `PAM70` and `PAM250`.
+
+        - BLOcks SUbstitution Matrix (BLOSUM) [2]_: A set of amino acid
+          substitution matrices, including `BLOSUM45`, `BLOSUM50`, `BLOSUM62`,
+          `BLOSUM80` and `BLOSUM90`.
+
+        References
+        ----------
+        .. [1] Dayhoff, M., Schwartz, R., & Orcutt, B. (1978). A model of
+           evolutionary change in proteins. Atlas of protein sequence and
+           structure, 5, 345-352.
+        .. [2] Henikoff, S., & Henikoff, J. G. (1992). Amino acid substitution
+           matrices from protein blocks. Proceedings of the National Academy of
+           Sciences, 89(22), 10915-10919.
+
+        Examples
+        --------
+        >>> from skbio import SubstitutionMatrix
+        >>> mat = SubstitutionMatrix.by_name('BLOSUM62')
+        >>> len(mat.alphabet)
+        24
+        >>> mat['M', 'K']
+        -1.0
+
+        """
+        try:
+            return named_substitution_matrices[name]
+        except KeyError:
+            raise ValueError(f'Substitution matrix "{name}" does not exist.')
+
+    @classonlymethod
+    @experimental(as_of='0.5.10')
+    def get_names(cls):
+        """List names of pre-defined substitution matrices.
+
+        Returns
+        -------
+        list of str
+            Names of pre-defined substitution matrices.
+
+        See Also
+        --------
+        by_name
+        """
+        return list(named_substitution_matrices.keys())
+
+
+def _matrix_to_vector(mat):
+    """Convert a square matrix into a flattened vector consisting of the upper
+    triangle and the diagonal.
+    """
+    return mat[np.triu_indices(len(mat))]
+
+
+def _vector_to_matrix(vec):
+    """Convert a vector consisting of the upper triangle and the diagonal of a
+    matrix to square form.
+    """
+    n = int((np.sqrt(1 + 8 * len(vec)) - 1) / 2)
+    mat = np.zeros((n, n))
+    mat[np.triu_indices(n)] = vec
+    return mat + np.triu(mat, k=1).T
+
+
+# Defined according to the matrices hosted at the NCBI FTP server:
+# https://ftp.ncbi.nlm.nih.gov/blast/matrices/
+named_substitution_matrices = {
```

Comment:
I like the idea of lazy loading. I am not familiar with the best approaches to achieve this. Some other packages simply store the matrices as external files. I wonder if there is a better approach? It may be the greatest if you can show an example.

Reply:
Something like the following, and would preserve dict-like interaction?

```python
class _LazySubstitutionMatrix:
    def __init__(self, name, path):
        self.name = name
        self.path = path
        self._data = None
    
    @property
    def data(self):
        if self._data is None:
            self._data = _some_mechanism_to_load_data(self.path)
        return self._data

class _SubstitutionMatrices:
    _matrices = (_LazySubstitutionMatrix('NUC.4.4', 'file://foo/bar'),
                 _LazySubstitutionMatrix('PAM30', http://foo/bar'), 
                 ...)
    def __init__(self):
        self._lazy_data = {m.name: m for m in self._matrices}
     
    def __get__(self, k):
        return self._lazy_data[k].data

    def keys(self):
        return sorted(self._lazy_data)

    def values(self):
        return [self[k] for k in self.keys()]

    def items(self):
        return [(k, self[k]) for k in self.keys()]

named_substitution_matrices = _SubstitutionMatrices()
```

Reply:
Relevant to the other comment. Will put this on hold pending the discussion on external data.

Reply:
PS @wasade : If we directly type data in `_LazySubstitutionMatrix` (instead of external files), will it still suffice lazy loading?

Reply:
You could defer the calls to `_vector_to_matrix` for example but bigger bang may be until these can be pulled from an authoritative source

Reply:
I envision that skbio should have a local data repository that is distributed with the package, plus a remote server that stores larger datasets. Analogs are scikit-learn's small, local datasets and pyTorch's big, remote datasets.

For substitution matrices specifically, the [NCBI repo](https://ftp.ncbi.nlm.nih.gov/blast/matrices/) has many common matrices, but not all. These are only for alignments. Meanwhile, matrices for tree building are not there. Examples are JTT, WAG, LG, etc. In comparison, BioPython has some matrices that are not found in NCBI. For skbio, I think that having a local repo of most common matrices (like the ones I already typed) and a remote repo of extended ones may be the way to go.

Reply:
Sounds good! Could an issue be opened to capture this? It could refer back to the discussion here